### ISSUE TITLE
Release 4.3.0 - Release Candidate 3 - AWS Module testing

### ISSUE THREAD

**Issue by CarlosRS9** [wazuh/wazuh#11775]
# AWS module testing

## Modules tests information
|                                      |                                                |
|--------------------------------------|------------------------------------------------|
| **Main module tests issue**          | #11862   |
| **Version**                          | 4.3.0                                          |
| **Release candidate #**              | RC3                                            |
| **Tag**                              | - |
| **Previous modules tests issue**     | #11427                |

## Test report procedure

All test results must have one the following statuses: 
|                 |                                                                         |
|-----------------|-------------------------------------------------------------------------|
| :green_circle:  | The test passed successfully. 										    |
| :yellow_circle: | The test passed but some enhancements need to be applied to the module. |
| :red_circle:    | The test failed.                                                        |
| :white_circle:  | The test does not apply.                                                |

Any failing test must be properly addressed with a new issue, detailing the error and the possible cause. It must be included in the [Release 4.3.0 project](https://github.com/orgs/wazuh/projects/4/) for triage.

Any expected fail or skipped test must have an issue justifying the reason. All auditors must validate the justification for an expected fail or skipped test.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

Every test will be performed on a Manager unless otherwise specified.

## Conclusions

| Status | Module | Notes |
|--------|--------|-------|
| :yellow_circle: | CloudWatch Logs       |`RC1` #10663: There is no support for reparsing in that module yet  |
| :red_circle: | Umbrella | #11899: The module is misfunctioning when executing it for the first time without a `only_logs_after` value |
| :red_circle: | Inspector | #11930: The module is not properly using the `only_logs_after` parameter<br>#11561: The module is not showing if the reparse option is being set although it works properly |
## Test results

The tests performed emphasize the available parameters as they have been modified or are affected by changes applied in this release:

| Bucket/Service type                                                                                  | only_logs_after | path           | path_suffix    | reparse           | discard_regex  | Issue ref. |
|------------------------------------------------------------------------------------------------------|-----------------|----------------|----------------|-------------------|----------------|------------|
| [VPC](https://github.com/wazuh/wazuh/issues/11775#issuecomment-1010070513)                 | :green_circle:  | :white_circle: | :white_circle: | :green_circle:    | :white_circle: | - |
| [Config](https://github.com/wazuh/wazuh/issues/11775#issuecomment-1010070648)              | :green_circle:  | :white_circle: | :white_circle: | :green_circle:    | :white_circle: | - |
| [CloudWatch Logs](https://github.com/wazuh/wazuh/issues/11775#issuecomment-1010070746)     | :green_circle:  | :white_circle: | :white_circle: | :yellow_circle:   | :white_circle: | #10663 |
| [Umbrella](https://github.com/wazuh/wazuh/issues/11775#issuecomment-1010070840)            | :red_circle:     | :green_circle:       | :green_circle:        | :green_circle:           | :green_circle:        | #11899 |
| [Inspector](https://github.com/wazuh/wazuh/issues/11775#issuecomment-1013058557)           | :red_circle:         | :white_circle: | :white_circle: | :yellow_circle:  | :white_circle: | #11930, #11561 |

The following bucket/service types were successfully tested in RC2 and they worked as expected in the previous RC2, so only [basic testing](https://github.com/wazuh/wazuh/issues/11775#issuecomment-1010070912) was performed this time. More detailed tests can be found in #11428

| Bucket/Service type | Test result | Issue ref. |
|---------------------|-------------|------------|
| CloudTrail          | :green_circle:     | -          |
| ALB                 | :green_circle:     | -          |
| CLB                 | :green_circle:     | -          |
| NLB                 | :green_circle:     | -          |
| KMS                 | :green_circle:     | -          |
| Macie               | :green_circle:     | -          |
| Trusted Advisor     | :green_circle:     | -          |
| GuardDuty           | :green_circle:     | -          |
| WAF                 | :green_circle:     | -          |
| Server Access       | :green_circle:     | -          |

The following test was performed to ensure the module worked as expected in a Wazuh Agent installation:

|   | Test result | Issue ref. |
|---|-------------| - |
| Test the module in an agent | :green_circle: | - |

## Auditors validation
The definition of done for this one is the validation of the conclusions and the test results from all auditors.

All checks from below must be accepted in order to close this issue.

- [x] @davidjiglesias
- [x] @CarlosRS9
- [x] @noise-kngdm

**Comment by CarlosRS9** [comment]
# VPCFlow

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events were sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Test                  | Status | Issue ref. |
|-----------------------|--------|------------|
| Test only_logs_after  | | |
| Using reparse         | | |

## Test details
### Test only_logs_after

Remove the `/var/ossec/wodles/aws/s3_cloudtrail.db` file and run the module using the following command to obtain every log from the VPCFlow bucket for the given date.

#### First execution without specifying an only_logs_after value
Upload one log file to the bucket with the date of execution. Then execute the command. It should use as marker an string with the date of execution and collect the uploaded log.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### First execution specifying an only_logs_after value set in the past
Remove the database file and run the command the using `2021-NOV-09` date as `only_logs_after`. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -s 2021-NOV-09 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

Keep the database file and run the same command again to ensure no duplicate logs have been processed.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -s 2021-NOV-09 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


#### Execute the command again with an only_logs_after value set earlier in the past

Run the module again, keeping the database file, using the `2021-NOV-05` date as `only_logs_after`. It should use as marker an string with the date of execution and not consider the `only_logs_after` value.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -s 2021-NOV-05 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


#### Execute the module having a last key set and no only_logs_after value
Without removing the database, execute `sqlite3 /var/ossec/wodles/aws/s3_cloudtrail.db 'delete from vpcflow where created_date in (select max(created_date) from vpcflow);'`. This command will remove the more recents logs. Then execute the module without specifying an `only_logs_after` value. It should only collect the previously uploaded log.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


#### Execute the module having an only_logs_after later than the last key
The module should only fetch the logs corresponding to the days set by the only_logs_after date onwards, and it should skip the logs in between. 2 logs should be processed.
Keeping the database file, execute `sqlite3 /var/ossec/wodles/aws/s3_cloudtrail.db 'delete from vpcflow where created_date > 20211109;'`. 

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -p dev -d2 -s YYYY-MMM-DD
```
Being `YYYY-MMM-DD` the data corresponding to the date of execution.

<details><summary>Command output</summary>

</details>

##### Conclusions

#### Execute the module having an only_logs_after earlier than the last key
Without removing the database, execute `sqlite3 /var/ossec/wodles/aws/s3_cloudtrail.db 'delete from vpcflow where created_date not in (select min(created_date) from vpcflow);'`. This command will remove all the logs but the older ones. The execution of the module must result in collecting 2 logs, using as marker the last key.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -p dev -d2 -s 2021-Nov-05
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Using reparse

Remove the `/var/ossec/wodles/aws/s3_cloudtrail.db` file and run the module using the following command to obtain every log from the VPCFlow bucket for the given date.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -s 2021-NOV-09 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

Run the module again, keeping the database file, using the same `only_logs_after` value and the `reparse` option. The same logs should be processed again, resulting in duplicated events.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -s 2021-NOV-09 --reparse -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

#### Conclusions

**Comment by CarlosRS9** [comment]
# Config

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events were sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Test | Status | Issue ref. |
|---|---|---|
| Test only_logs_after  |  | |
| Using reparse         |  | |

### Conclusions

## Test details
### Test only_logs_after

Remove the `/var/ossec/wodles/aws/s3_cloudtrail.db` file and run the module using the following command to obtain every log from the Config bucket for the given date.

#### First execution without specifying an only_logs_after value
Upload one log file to the bucket with the date of execution. Then execute the command. It should use as marker an string with the date of execution and collect the uploaded log.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### First execution specifying an only_logs_after value set in the past
Remove the database file and run the command the using `2021-NOV-22` date as `only_logs_after`. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -s 2021-NOV-22 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

Keep the database file and run the same command again to ensure no duplicate logs have been processed.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -s 2021-NOV-22 -p dev -d2
```

<details>
    <summary>Command output</summary>


</details>

##### Conclusions

#### Execute the command again with an only_logs_after value set earlier in the past

Run the module again, keeping the database file, using the `2021-NOV-06` date as `only_logs_after`. It should use as marker an string with the date of execution and not consider the `only_logs_after` value.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -s 2021-NOV-06 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### Execute the module having a last key set and no only_logs_after value
Without removing the database, execute `sqlite3 /var/ossec/wodles/aws/s3_cloudtrail.db 'delete from config where created_date in (select max(created_date) from config);'`. This command will remove the more recents logs. Then execute the module without specifying an `only_logs_after` value. It should only collect the previously uploaded log.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### Execute the module having an only_logs_after later than the last key
The module should only fetch the logs corresponding to the days set by the only_logs_after date onwards, and it should skip the logs in between. 1 log should be processed.
Keeping the database file, execute `sqlite3 /var/ossec/wodles/aws/s3_cloudtrail.db 'delete from config where created_date > 20211122;'`. 

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -p dev -d2 -s YYYY-MMM-DD
```
Being `YYYY-MMM-DD` the data corresponding to the date of execution.

<details><summary>Command output</summary>


</details>

##### Conclusions

#### Execute the module having an only_logs_after earlier than the last key
Without removing the database, execute `sqlite3 /var/ossec/wodles/aws/s3_cloudtrail.db 'delete from config where created_date not in (select min(created_date) from config);'`. This command will remove all the logs but the older ones. The execution of the module must result in collecting 2 logs, using as marker the last key.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -p dev -d2 -s 2021-Oct-01
```

<details>
    <summary>Command output</summary>


</details>

##### Conclusions



### Using reparse

Remove the `/var/ossec/wodles/aws/s3_cloudtrail.db` file and run the module using the following command to obtain every log from the Config bucket for the given date.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -s 2021-NOV-21 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

Run the module again, keeping the database file, using the same `only_logs_after` value and the `reparse` option. The same logs should be processed again, resulting in duplicated events.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -s 2021-NOV-21 --reparse -p dev -d2
```

<details>
    <summary>Command output</summary>


#### Conclusions

**Comment by CarlosRS9** [comment]
# CloudWatch Logs

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events was sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Test | Status | Issue ref. |
|---|---|---|
| Using only_logs_after | |  |
| Using reparse | | |


## Test details
Before starting these tests, a series of Cloudwatch Logs test events in a period of 14 days in the past until today must be uploaded to the `4_3_rc2_test`'s `test_stream` log stream.

### Test only_logs_after

Remove the `/var/ossec/wodles/aws/aws_services.db` file and run the module using the following command to obtain every log from the Cloudwatch Logs log group for the given date.

#### First execution without specifying an only_logs_after value

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_3_rc2_test -d 2 -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### First execution specifying an only_logs_after value set in the past
Remove the database file and run the command specifying a date set 3 days earlier in the past. All the logs between the date set by the `only_logs_after` parameter and the present day must be collected.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_3_rc2_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

</details>

Keep the database file and run the same command again to ensure no duplicate logs are processed. Use the same `only_logs_after` parameter.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_3_rc2_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### Execute the command again with an only_logs_after value set earlier in the past

Run the module again, keeping the database file, using the date set 14 days earlier in the past than the execution time. The module should collect all logs between that date and the first log collected and skip the rest to avoid processing duplicates.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_3_rc2_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### Execute the module having a last key set and no only_logs_after value
Without removing the database file, execute `sqlite3 /var/ossec/wodles/aws/aws_services.db 'UPDATE cloudwatch_logs SET end_time = XXXXXXXXXXXXX, next_token=null;'`, being `XXXXXXXXXXXXX` the timestamp in milliseconds corresponding to the fourth log.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_3_rc2_test -d 2 -p dev -r us-east-1
```
The module should collect all the logs starting from the `Test log 5`.

<details>
    <summary>Command output</summary>


</details>

##### Conclusions

#### Execute the module having an only_logs_after later than the last key
Keeping the database file, execute `sqlite3 /var/ossec/wodles/aws/aws_services.db 'UPDATE cloudwatch_logs SET end_time = XXXXXXXXXXXXX next_token=null;'`, being `XXXXXXXXXXXXX` the timestamp in milliseconds corresponding to the fourth log.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_3_rc2_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```
Being `YYYY-MMM-DD` the date corresponding to the day following the execution date. No log should be collected.

<details>
    <summary>Command output</summary>


</details>

##### Conclusions

#### Execute the module having an only_logs_after earlier than the last key
Keeping the database file, execute `sqlite3 /var/ossec/wodles/aws/aws_services.db 'UPDATE cloudwatch_logs SET end_time = XXXXXXXXXXXXX, start_time = YYYYYYYYYYYYY, next_token=null;'`, being `XXXXXXXXXXXXX` the timestamp in milliseconds corresponding to the sixth log, and `YYYYYYYYYYYYY` the timestamp in milliseconds corresponding to the fourth log. Then run the module specifying the day before the date of execution as `only_logs_after`. The module should collect all logs in the log stream except for those with a timestamp between `YYYYYYYYYYYYY` and `XXXXXXXXXXXXX`.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_3_rc2_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

### Using reparse

This option is not available for the CloudWatch Logs integration. It will be added in #10663.

**Comment by CarlosRS9** [comment]
# Umbrella

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events were sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Test | Status | Issue ref. |
|---|---|---|
| Test only_logs_after |  | |
| Using path            |  | |
| Using reparse         |  | |
| Using discard_regex   |  | |
| Test iplogs   |   | |
| Test dnslogs   |   | |
| Test proxylogs   |   | |


## Test details
### Test only_logs_after
Since the three types of Cisco Umbrella buckets use the same underlying logic, we will only need to test that the `only_logs_after` parameter for one of them.

Remove the `/var/ossec/wodles/aws/s3_cloudtrail.db` file and run the module using the following command to obtain every log from the Umbrella bucket for the given date.

#### First execution without specifying an only_logs_after value
Upload one log file to the bucket with the date of execution. Then execute the command. It should use as marker an string with the date of execution and collect the uploaded log.

Use the command
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l iplogs -t cisco_umbrella -p dev -d2
```

<details>
    <summary>Command output</summary>


</details>

##### Conclusions

**Comment by CarlosRS9** [comment]
# Other bucket types and services

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events were sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Bucket/Service type | Test result | Issue ref. |
|---------------------|-------------|------------|
| CloudTrail          | Pending     | -          |
| ALB                 | Pending     | -          |
| CLB                 | Pending     | -          |
| NLB                 | Pending     | -          |
| KMS                 | Pending     | -          |
| Macie               | Pending     | -          |
| Trusted Advisor     | Pending     | -          |
| GuardDuty           | Pending     | -          |
| WAF                 | Pending     | -          |
| Server Access       | Pending     | -          |

## Test details


### Test Cloudtrail

Remove the database file and run the command the using 2021-NOV-23 date as only_logs_after. 6 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-cloudtrail -t cloudtrail -s 2021-NOV-23 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

### Test ALB

Remove the database file and run the command the using 2020-Dec-21 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-alb -t alb -s 2020-Dec-21 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test CLB

Remove the database file and run the command the using 2021-NOV-11 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-clb -t clb -s 2021-NOV-11 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

### Test NLB

Remove the database file and run the command the using 2021-NOV-11 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-nlb -t nlb -s 2020-NOV-21 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test KMS

Remove the database file and run the command the using 2021-NOV-19 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-kms -t custom -s 2021-NOV-19 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test Macie

Remove the database file and run the command the using 2021-NOV-11 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-macie -t custom -s 2021-NOV-11 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test Trusted Advisor

Remove the database file and run the command the using 2021-Dec-15 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-trusted-advisor -t custom -s 2021-Dec-15 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions




### Test Guard Duty

Remove the database file and run the command the using 2021-Jun-22 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-guardduty -t custom -s 2021-Jun-22 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test WAF

Remove the database file and run the command the using 2021-Dec-22 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-waf -t custom -s 2021-Sep-17 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test Server Access

Remove the database file and run the command the using 2021-NOV-12 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-access -t server_access -s 2021-NOV-12 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

**Comment by CarlosRS9** [comment]
# Agent testing

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events were sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Test | Status | Issue ref. |
|---|---|---|
| Test the module in an agent |  | - |



## Test the module in an agent

Deploy an agent and install the required dependencies described in the official Wazuh documentation, then run the module adding the following configuration to the ossec.conf:

<details>
    <summary>Configuration</summary>

    <wodle name="aws-s3">
        <disabled>no</disabled>
        <remove_from_bucket>no</remove_from_bucket>
        <interval>10m</interval>
        <run_on_start>yes</run_on_start>
        <skip_on_error>no</skip_on_error>

        <bucket type="cloudtrail">
            <aws_profile>dev</aws_profile>
            <name>wazuh-aws-wodle-cloudtrail</name>
            <only_logs_after>2021-NOV-21</only_logs_after>
        </bucket>

        <bucket type="custom">
            <aws_profile>dev</aws_profile>
            <name>wazuh-aws-wodle-macie</name>
            <only_logs_after>2021-NOV-10</only_logs_after>
        </bucket>

        <service type="cloudwatchlogs">
            <aws_profile>dev</aws_profile>
            <aws_log_groups>framework-test-apache</aws_log_groups>
            <only_logs_after>2021-AUG-20</only_logs_after>
            <regions>us-east-1</regions>
        </service>
    </wodle>
</details>

<details>
    <summary>Output</summary>

</details>

#### Conclusions

**Comment by noise-kngdm** [comment]
# Inspector

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events was sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Test | Status | Issue ref. |
|---|---|---|
| Using only_logs_after | |  |
| Using reparse | | |


## Test details
Before starting these tests, at least three assessments runs must have been run in three different dates.

### Test only_logs_after

Remove the `/var/ossec/wodles/aws/aws_services.db` file and run the module using the following command to obtain every finding log group for the given date.

#### First execution without specifying an only_logs_after value

Use the command: 
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -p dev -r us-east-1
```

Unless a new assessment run was executed in the same date, no log should be collected.

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### First execution specifying an only_logs_after value set in the past
Remove the database file and run the command specifying the date of the last assessment run. All events from that date onwards should be collected.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

</details>

Keep the database file and run the same command again to ensure no duplicate logs are processed. Use the same `only_logs_after` parameter.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### Execute the command again with an only_logs_after value set earlier in the past

Run the module again, keeping the database file, using the date set for the first assessment run registered. The module should process no new log.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### Execute the module having a last key set and no only_logs_after value
Remove the database and execute the command `/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -p dev -r us-east-1 -s 2022-Jan-03`, which is the date of the first assessment run.
Then execute `sqlite3 ./aws_services.db 'delete from aws_services where scan_date != (SELECT min(scan_date) from aws_services);'`. This will simulate an environment which has only collected the logs corresponding to the first assessment run.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -p dev -r us-east-1
```
The module should collect all the logs starting from the `2022-Jan-03` without processing duplicates.

<details>
    <summary>Command output</summary>


</details>

##### Conclusions

#### Execute the module having an only_logs_after later than the last key
Keeping the database file, execute `sqlite3 ./aws_services.db 'delete from aws_services where scan_date != (SELECT min(scan_date) from aws_services);'`.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```
Being `YYYY-MMM-DD` the date corresponding to the day following the execution date. No log should be collected.

<details>
    <summary>Command output</summary>


</details>

##### Conclusions

#### Execute the module having an only_logs_after earlier than the last key
Keeping the database file, execute `sqlite3 ./aws_services.db 'delete from aws_services where scan_date != (SELECT min(scan_date) from aws_services);'`.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s 2021-Dec-01 -p dev -r us-east-1
```

It should only collect the logs starting from the first assessment run without processing it again.

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

### Using reparse
Without removing the databases file, execute the following command:

```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s 2021-Jan-03 -p dev -r us-east-1 --reparse
```

<details>
    <summary>Command output</summary>

</details>

#### Conclusions
